### PR TITLE
Spectroelectrochemistry

### PR DESCRIPTION
**Pull request by ScottSoren**
Hi @Caiwu-L , 
Thanks for agreeing to review this PR!

This is the code behind the spectroelectrochemistry interface you helped me develop, demonstrated here: https://github.com/ixdat/tutorials/blob/main/spectroelectrochemistry/spectroelectrochemistry_demo.ipynb

I've also included a first iteration of the Spectroelectrochemistry documentation in this. To view that in .html form, you will need to follow the [sphinx instructions here](https://github.com/ixdat/ixdat/blob/6fc3bbbddff0ceff8270431c43ba1ca21c3ae0cb/TOOLS.rst#L76).

I'd appreciate it if you could request changes:

- anywhere the code is unclear (probably needs comments!)
- anywhere the code doesn't look like it does what you think it should (needs to be improved!)
- any missing docstrings for classes, functions, and methods
- anywhere the documentation is hard to read (needs to be fixed!).

The addition is a lot of lines of code because the SEC readers, exporters, and plotters were complex, and because I developed `Spectrum` and `SpectrumSeries` for this. On the latter point, I'm not sure it's the best structure now, so will appreciate thoughts on that! I tried to do it all in ways that will require the least additional code for e.g. EC-FTIR, etc. The most important module for your checking and full understanding is of course `spectroelectrochemistry`. 

Thanks a lot, and looking forward to a good review! :D

### CODE REVIEW

File: src/ixdat/readers/msrh_sec.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        cls=None,
+    ):
+        """Read potential-dep. SEC data from 3 csv's to return a SpectroECMeasurement
+
+        The function is well-commented so take a look at the source
+
+        Args:
+            path_to_file (Path or str): The full path to the file containing the
+                spectra data. This file has voltage in the first row, and a first
+                column with an arbitrary counter which has to be replaced by wavelength.
```
Comment (Caiwu-L) [inline src/ixdat/readers/msrh_sec.py:29]: 'arbitrary counter' change to 'pixels number that related to wavelength'.

Comment (Caiwu-L) [inline src/ixdat/readers/msrh_sec.py:29]: 'arbitrary counter' change to 'pixels number that related to wavelength'?

Comment (Caiwu-L) [inline src/ixdat/readers/msrh_sec.py:29]: 'arbitrary counter' change to 'pixels number that related to wavelength'?

File: src/ixdat/readers/msrh_sec.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+        # us pandas to load the data from the csv files into dataframes:
+        sec_df = pd.read_csv(path_to_file)
+        # ^ Note the first row, containing potential, will become the keys. The first
+        #   column, containing an arbitrary counter, is included in the data.
+        ref_df = pd.read_csv(path_to_ref_spec_file, names=["wavelength", "counts"])
+        jv_df = pd.read_csv(path_to_V_J_file, names=["v", "j"])
+
+        # The spectra need (i) the first colum with the arbitrary counter to be
+        #    discarded and (ii) axes switched so that wavelength is the inner axis
+        #    (axis=1). The outer axis (axis=0) then spans potential or, eq., time:
```
Comment (Caiwu-L) [inline src/ixdat/readers/msrh_sec.py:60]: what is inner axis and outer axis?

File: src/ixdat/readers/msrh_sec.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+        wl = wl[excess_wl_points:]
+        ref_signal = ref_df["counts"].to_numpy()[excess_wl_points:]
+
+        # Now we're ready to define all the spectrum DataSeries:
+
+        # wavelength is independent variable --> simple DataSeries
+        wl_series = DataSeries("wavelength / [nm]", "nm", wl)
+        # The reference spectrum spans a space defined by wavelength:
+        reference = Field(
+            "reference",
```
Comment (Caiwu-L) [inline src/ixdat/readers/msrh_sec.py:87]: name=''reference''
unit_name=''counts''

File: src/ixdat/readers/msrh_sec.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        #    discarded and (ii) axes switched so that wavelength is the inner axis
+        #    (axis=1). The outer axis (axis=0) then spans potential or, eq., time:
+        spectra = sec_df.to_numpy()[:, 1:].swapaxes(0, 1)
+        # The potential comes from the keys of that data, discarding the first column:
+        v = np.array([float(key) for key in sec_df.keys()])[1:]
+        # We get time from this potential and the scan rate, with a helper function:
+        t = calc_t_using_scan_rate(v, dvdt=scan_rate * 1e-3)
+        # If they didn't provide a tstamp, we have to prompt for it.
+        tstamp = tstamp or prompt_for_tstamp(path_to_file)
+        # Ready to define the measurement's TimeSeries:
```
Comment (Caiwu-L) [inline src/ixdat/readers/msrh_sec.py:68]: I like this kind of comments

File: src/ixdat/readers/msrh_sec.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+            data=spectra,
+        )
+
+        # Now we process the current and potential:
+        v_0 = jv_df["v"].to_numpy()  # ... but we'll actually use v from the sec data
+        j_0 = jv_df["j"].to_numpy() * 1e3  # 1e3 converts [A] to [mA]
+        if v_0[0] > v_0[-1]:  # Need the potential in the EC file to be increasing:
+            v_0 = np.flip(v_0)
+            j_0 = np.flip(j_0)
+        # Since the "real" potential is in the sec data, we need to interpolate the
```
Comment (Caiwu-L) [inline src/ixdat/readers/msrh_sec.py:106]: In the current version of data form labview, I found the potential in SEC is always one point less than that in J-V, so it is fine, but very good to have this for other cases.

File: src/ixdat/readers/msrh_sec.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+        #   current onto it:
+        j = np.interp(v, v_0, j_0)
+        # and now we're ready to define the electrochemical DataSeries:
+        v_series = ValueSeries("raw potential / [V]", "V", v, tseries=tseries)
```
Comment (Caiwu-L) [inline src/ixdat/readers/msrh_sec.py:110]: good to have the name of each kwargs?

File: src/ixdat/readers/msrh_sec.py [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+        series_list = [
+            tseries,
+            v_series,
+            j_series,
+            wl_series,
+            reference,
+            spectra,
+        ]
+
+        # Figure out which measurement class to return. Use S-EC unless this read
```
Comment (Caiwu-L) [inline src/ixdat/readers/msrh_sec.py:123]: can't follow this part, could you add more comments here?

File: src/ixdat/readers/msrh_sec.py [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+        The function is well-commented so take a look at the source
+
+        Args:
+            path_to_file (Path or str): The full path to the file containing the
+                spectra data. This file has voltage in the first row, and a first
+                column with an arbitrary counter which has to be replaced by wavelength.
+            path_to_ref_spec_file (Path or str): The full path to the file containing
+                the wavelenth data, together usually with the adsorption-free spectrum.
+                The length of the columns should be the same as in the spectrum data
+                but in practice is a few points longer. The excess points at the ends
```
Comment (Caiwu-L) [inline src/ixdat/readers/msrh_sec.py:33]: seems the points at the begining is discarded from line 79 ?

File: src/ixdat/techniques/spectroelectrochemistry.py [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+    def spectrum_series(self):
+        """The SpectrumSeries that is the spectra of the SEC Measurement"""
+        return SpectrumSeries.from_field(
+            self.spectra,
+            tstamp=self.tstamp,
+            name=self.name + " spectra",
+        )
+
+    @property
+    def wavelength(self):
```
Comment (Caiwu-L) [inline src/ixdat/techniques/spectroelectrochemistry.py:69]: what is the difference for wavelength and wl, why need one dataseries and one np array?

File: src/ixdat/techniques/spectroelectrochemistry.py [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+            t_ref (float): The time at which to get the reference spectrum
+            index_ref (int): The index of the reference spectrum
+        Returns ValueSeries: The dOD value of the spectrum at wl.
+        """
+        if V_ref or t_ref or index_ref:
+            spectrum_ref = self.get_spectrum(V=V_ref, t=t_ref, index=index_ref)
+        else:
+            spectrum_ref = self.reference_spectrum
+        x = self.wl
+        if width:  # averaging
```
Comment (Caiwu-L) [inline src/ixdat/techniques/spectroelectrochemistry.py:230]: Can't understand here, if width is given, then average according to with, else, we average according the default width 10. why here else:  interpolation?

File: src/ixdat/techniques/spectroelectrochemistry.py [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+
+        Args:
+            spectrum (Spectrum or str): If a Spectrum is given, it becomes the reference
+                spectrum. The string "reference" can be given to make the reference
+                spectrum become (via the reference_spectrum property) one that the
+                measurement was loaded with (evt. for definition of wavelengths).
+            t_ref (float): The time (with respect to self.tstamp) to use as the
+                reference spectrum
+            V_ref (float): The potential to use as the reference spectrum. This will
+                only work if the potential is monotonically increasing.
```
Comment (Caiwu-L) [inline src/ixdat/techniques/spectroelectrochemistry.py:44]: V also need to increase with constant interval?

File: src/ixdat/techniques/spectroelectrochemistry.py [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+        )
+        return Spectrum.from_field(field)
+
+    def track_wavelength(self, wl, width=10, V_ref=None, t_ref=None, index_ref=None):
+        """Return and cache a ValueSeries for the dOD for a specific wavelength.
+
+        The cacheing adds wl_str to the SECMeasurement's data series, where
+            wl_str = "w" + int(wl)
+            This is dOD. The raw is also added as wl_str + "_raw".
+        So, to get the raw counts for a specific wavelength, call this function as
```
Comment (Caiwu-L) [inline src/ixdat/techniques/spectroelectrochemistry.py:210]: call as.... miss some words here?

File: src/ixdat/plotters/spectrum_plotter.py [hunk 13]
```diff
@@ -1,15 +1,218 @@
...
+        """
+        spectrum_series = spectrum_series or self.spectrum_series
+        field = field or spectrum_series.field
+
+        xseries = field.axes_series[1]
+        x = xseries.data
+        tseries = field.axes_series[0]
+
+        v_name = vs
+        if vs in ("t", tseries.tseries.name):
```
Comment (Caiwu-L) [inline src/ixdat/plotters/spectrum_plotter.py:118]: tseries.tseries.name?

File: src/ixdat/plotters/sec_plotter.py [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+        cmap_name="inferno",
+        make_colorbar=False,
+        **kwargs,
+    ):
+        """Plot an SECMeasurement in two panels with potential as x-asis.
+
+        The top panel is a heat plot with wavelength on y-axis and color representing
+        spectrum. At most one of V_ref and t_ref should be given, and if neither are
+        given the measurement's default reference_spectrum is used to calculate the
+        optical density.
```
Comment (Caiwu-L) [inline src/ixdat/plotters/sec_plotter.py:154]: optical density could change to dO.D., to be consistent through all the comments and figures

File: src/ixdat/plotters/sec_plotter.py [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+        if not axes:
+            axes = self.new_two_panel_axes(n_bottom=2)
+        for wl_str in wavelengths:
+            x = float(wl_str[1:])
+            try:
+                t, y = measurement.grab(wl_str, tspan=tspan)
+            except SeriesNotFoundError:
+                measurement.track_wavelength(x)
+                t, y = measurement.grab(wl_str, tspan=tspan)
+            axes[0].plot(t, y, color=cmap(norm(x)), label=wl_str)
```
Comment (Caiwu-L) [inline src/ixdat/plotters/sec_plotter.py:230]: No label for Y axis in the top figure currently

File: src/ixdat/plotters/sec_plotter.py [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+        axes[0].legend()
+
+        self.ec_plotter.plot_measurement(
+            measurement=measurement, axes=axes[1:], tspan=tspan, **kwargs
+        )
+
+    def plot_wavelengths_vs_potential(
```
Comment (Caiwu-L) [inline src/ixdat/plotters/sec_plotter.py:237]: Also could add a label for Y axis in the top figure.

File: src/ixdat/techniques/spectroelectrochemistry.py (hunk 10 above)
Comment (ScottSoren) [reply src/ixdat/techniques/spectroelectrochemistry.py:230]: Oh, right. The user would have to write `width=False` or `width=0` to force it to interpolate. I have added this now to the docstring.

**Review by ScottSoren — COMMENTED** [review]
@Caiwu-L,
Thanks for a great review! You caught some really subtle mistakes. 
I've corrected and added documentation to address each of your points, and replied to the comments.
I also fixed the bug in exporting and re-reading S-EC data. The bug was that `set_reference_spectrum` was resulting in a reference spectrum with `name=spectra` that clashed with the main 2-D data set that also has `name=spectra`. Now I've ensured that the reference spectrum gets a different name.
 I'll push the updated code now.

File: src/ixdat/readers/msrh_sec.py (hunk 6 above)
Comment (Caiwu-L) [reply src/ixdat/readers/msrh_sec.py:110]: Agree :)

**Pull request merged**